lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,420 +25,540 @@
+        if (isa<RegOp, RegResetOp, SubaccessOp, SubfieldOp, SubindexOp,
+                RefSubOp>(def))
+          return;
+
+    auto dstValType = getBaseType(dstVal.getType());
+    auto srcValType = getBaseType(srcVal.getType());
+
+    // Handle Properties.
+    if (!(srcValType && dstValType))
+      return addReachingDef({dstVal, 0}, {srcVal, 0});
+
+    auto addDef = [&](FieldRef dst, FieldRef src) {
+      // If the dstVal and srcVal are aggregate types, then record the dataflow
+      // between each individual ground type. This is equivalent to flattening
+      // the type to ensure all the contained FieldRefs are also recorded.
+      if (dstValType && !dstValType.isGround())
+        walkGroundTypes(dstValType, [&](uint64_t dstIndex, FIRRTLBaseType t) {
+          // Handle the case when the dst and src are not of the same type.
+          // For each dst ground type, and for each src ground type.
+          if (srcValType == dstValType)
+            addReachingDef(dst.getSubField(dstIndex),
+                           src.getSubField(dstIndex));
+          else if (srcValType && !srcValType.isGround())
+            walkGroundTypes(srcValType,
+                            [&](uint64_t srcIndex, FIRRTLBaseType t) {
+                              addReachingDef(dst.getSubField(dstIndex),
+                                             src.getSubField(srcIndex));
+                            });
+          // Else, the src is ground type.
+          else
+            addReachingDef(dst.getSubField(dstIndex), src);
+        });
+
+      addReachingDef(dst, src);
+    };
+
+    // Both the dstVal and srcVal, can refer to multiple FieldRefs, ensure that
+    // we capture the dataflow between each pair. Occurs when val result of
+    // subaccess op.
+
+    // Case 1: None of src and dst are fields of an aggregate. But they can be
+    // aggregate values.
+    if (dstIt == valToFieldRefs.end() && srcIt == valToFieldRefs.end())
+      return addDef({dstVal, 0}, {srcVal, 0});
+    // Case 2: Only src is the field of an aggregate. Get all the fields that
+    // the src refers to.
+    if (dstIt == valToFieldRefs.end() && srcIt != valToFieldRefs.end()) {
+      llvm::for_each(srcIt->getSecond(), [&](const FieldRef &srcField) {
+        addDef(FieldRef(dstVal, 0), srcField);
+      });
+      return;
+    }
+    // Case 3: Only dst is the field of an aggregate. Get all the fields that
+    // the dst refers to.
+    if (dstIt != valToFieldRefs.end() && srcIt == valToFieldRefs.end()) {
+      llvm::for_each(dstIt->getSecond(), [&](const FieldRef &dstField) {
+        addDef(dstField, FieldRef(srcVal, 0));
+      });
+      return;
+    }
+    // Case 4: Both src and dst are the fields of an aggregate. Get all the
+    // fields that they refers to.
+    llvm::for_each(srcIt->second, [&](const FieldRef &srcField) {
+      llvm::for_each(dstIt->second, [&](const FieldRef &dstField) {
+        addDef(dstField, srcField);
+      });
```

Comment:
I guess this is too pessimistic for bulk connections? All source elements are considered to connected to every dest elements so there might be false-positive combinatorial loops.
```
connect v1, v2
connect v2[0], 0
connect v2[1], v1[0]
```

I didn't actually test the example so maybe i'm wrong.